client/engine/engine.go

```diff
@@ -436,15 +436,25 @@ func (e *Engine) handlePaymentRequest(request PaymentRequest) error {
 	return e.executeSideEffects(se)
 }
 
-// executeSideEffects executes the SideEffects declared by cranking an Objective or handling a payment request.
-func (e *Engine) executeSideEffects(sideEffects protocols.SideEffects) error {
+// sendMessages sends out the messages and records the metrics.
+func (e *Engine) sendMessages(msgs []protocols.Message) {
 	defer e.metrics.RecordFunctionDuration()()
 
-	for _, message := range sideEffects.MessagesToSend {
+	for _, message := range msgs {
 		e.logMessage(message, Outgoing)
 		e.recordMessageMetrics(message)
 		e.msg.Send(message)
```

Comment:
This still looks like a blocking sequence to me. EG, where there are 3 messages (Alice wants to send to Irene, Ivan, and Bob), each send operation blocks the next.

Maybe:

``` golang
go func() {
  e.logMessage(message, Outgoing)
  e.recordMessageMetrics(message)
  e.msg.Send(message)
}()
```
so that the logging / metrics calls are closely bundled with the beginning of the send?

Reply:
You're right that within the `sendMessages` function that we're blocking on `e.msg.Send`. However I think that's ok as the main change in this PR to call `sendMessages` in a go-routine means the engine run loop no longer blocks on sending messages.

It's possible that calling `msg.Send` in a go-routine for each message might be slightly more efficient but I didn't notice that in the TTFP when I tried it in [62335ce](https://github.com/statechannels/go-nitro/pull/977/commits/62335ce507dd6c32afffd0b584c560fb18d63e8d). I think that makes sense as it's really the engine run loop that is the "hot path". By calling `sendMessages` in a go routine, we move sending messages off the "hot path", so it doesn't matter too much how fast `sendMessages` completes.

Reply:
The queue of messages is off the hot path for this particular client at this particular instant (ie, the client machine can get on immediately with processing new inputs), but the protocol at large (virtual fund, virtual defund) can be held up here.

I'm not exactly clear on the nature of the blocking inside of `msg.send`, but imagine where Alice wants to send a postFund signature to Irene, Ivan and Bob. The send to Bob won't happen until the sends to Irene and Ivan have completed, but this broadcast could be done in parallel.

Highlighted in @geoknee's draft PR, Bob is waiting for that post-fund message to finish the VF protocol.

(this would be expected to show in the numbers more in a multi-hop, since there are bigger queues of outgoing messages created there)

Reply:
> (this would be expected to show in the numbers more in a multi-hop, since there are bigger queues of outgoing messages created there) 

I've tried running a test with 5 hops and 20 concurrency.  The results seem similar between sending each message [in a go-routine ](http://34.168.92.245:3000/d/5OBBeW37k/time-to-first-payment?orgId=1&from=now-15m&to=now&var-runId=cdlioionr2gv4eib3akg&var-jobCount=20&var-testDuration=2m0s&var-hubs=5&var-payees=2&var-jitter=1&var-latency=10&var-payers=2&var-payeepayers=0&var-nitroVersion=v0.0.0-20221108220129-62335ce507dd) vs no [go-routine](http://34.168.92.245:3000/d/5OBBeW37k/time-to-first-payment?orgId=1&from=now-15m&to=now&var-runId=cdlimionr2gv4eib3ak0&var-jobCount=20&var-testDuration=2m0s&var-hubs=5&var-payees=2&var-jitter=1&var-latency=10&var-payers=2&var-payeepayers=0&var-nitroVersion=v0.0.0-20221107233712-45d9c84b5868) although that's far from an exhaustive test.

That said, it didn't seem to harm performance so I don't mind implementing that change!

Reply:
Hmm it seems that this change causes the tests to [fail](https://github.com/statechannels/go-nitro/actions/runs/3425382054/jobs/5706156123) for some reason 😞

Reply:
The testground run seemed to work just fine. Which made me think there could be an issue with the test messaging service receiving concurrent requests to send messages. Hence #985

Reply:
Based on standup discussion we're planning on just merging in with the original [change](https://github.com/statechannels/go-nitro/pull/977/commits/45d9c84b586810c8407e8d33d603fc2b1ed1db27). We will investigate this change and others when we look at performance in https://github.com/statechannels/go-nitro/issues/986

Reply:
Hmm it looks like even after this reverting back to the original change, we're still seeing [test failures](https://github.com/statechannels/go-nitro/actions/runs/3430904846/jobs/5718429968). This may be [related to the test message service](https://github.com/statechannels/go-nitro/pull/977#discussion_r1017819461) that @geoknee pointed out?

I am also curious if the test is just slow or if it's outright failing. Will investigate!

Reply:
I believe this is the [root cause](https://github.com/statechannels/go-nitro/pull/987) that is probably getting exposed by more concurrent message sending